core/dbt/events/types.py

```diff
@@ -1247,6 +1246,16 @@ def message(self) -> str:
         )
 
 
+class WarnStateTargetEqual(WarnLevel):
+    def code(self):
+        return "I072"
+
+    def message(self) -> str:
+        return yellow(
+            f"Warning: The state and target directories are the same: '{self.state_path}'"
```

Comment:
I like this message because its actionable/hints at how the user could prevent the warnings, but I do think it's missing something to explain why this is a warning in the first place. Could we work in "what could go wrong/unexpectedly because state and target directories are the same" to the message? 

My understanding from the issue is the potential for missing changes in state-based selection or retries that are not idempotent